core/store/src/db.rs

```diff
@@ -0,0 +1,187 @@
+use parking_lot::RwLock;
+use rocksdb::{
+    ColumnFamily, ColumnFamilyDescriptor, DBCompactionStyle, DBIterator, IteratorMode, Options,
+    ReadOptions, WriteBatch, DB,
+};
+use std::collections::HashMap;
+use std::io;
+
+fn other_io_err<E>(e: E) -> io::Error
+where
+    E: Into<Box<dyn std::error::Error + Send + Sync>>,
+{
+    io::Error::new(io::ErrorKind::Other, e)
+}
+
+pub struct DBTransaction {
+    pub ops: Vec<DBOp>,
+}
+
+pub enum DBOp {
+    Insert { col: usize, key: Vec<u8>, value: Vec<u8> },
+    Delete { col: usize, key: Vec<u8> },
+}
+
+impl DBTransaction {
+    pub fn put<K: AsRef<[u8]>, V: AsRef<[u8]>>(&mut self, col: usize, key: K, value: V) {
+        self.ops.push(DBOp::Insert {
+            col,
+            key: key.as_ref().to_owned(),
+            value: value.as_ref().to_owned(),
+        });
+    }
+
+    pub fn delete<K: AsRef<[u8]>>(&mut self, col: usize, key: K) {
+        self.ops.push(DBOp::Delete { col, key: key.as_ref().to_owned() });
+    }
+}
+
+pub struct RocksDB {
+    db: RwLock<Option<DB>>,
+    cfs: Vec<*const ColumnFamily>,
+    read_options: ReadOptions,
+}
+
+unsafe impl Send for RocksDB {}
+unsafe impl Sync for RocksDB {}
+
+pub struct TestDB {
+    db: RwLock<Option<Vec<HashMap<Vec<u8>, Vec<u8>>>>>,
+}
+
+pub trait Database: Sync + Send {
+    fn transaction(&self) -> DBTransaction {
+        DBTransaction { ops: Vec::new() }
+    }
+    fn get(&self, col: usize, key: &[u8]) -> Result<Option<Vec<u8>>, io::Error>;
+    fn iter<'a>(&'a self, column: usize) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a>;
+    fn write(&self, batch: DBTransaction) -> io::Result<()>;
+}
+
+impl RocksDB {
+    fn iterate(&self, col: usize) -> Option<DBIterator<'_>> {
+        match *self.db.read() {
+            Some(ref db) => unsafe {
+                let cf_handle = &*self.cfs[col];
+                let iterator =
+                    db.iterator_cf_opt(cf_handle, &self.read_options, IteratorMode::Start).unwrap();
+                Some(iterator)
+            },
+            _ => None,
+        }
+    }
+}
+
+impl Database for RocksDB {
+    fn get(&self, col: usize, key: &[u8]) -> Result<Option<Vec<u8>>, io::Error> {
+        match *self.db.read() {
+            Some(ref db) => unsafe {
+                Ok(db.get_cf_opt(&*self.cfs[col], key, &self.read_options).map_err(other_io_err)?)
+            },
+            None => Ok(None),
+        }
+    }
+
+    fn iter<'a>(&'a self, col: usize) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+        Box::new(RocksDB::iterate(self, col).into_iter().flat_map(|i| i))
+    }
+
+    fn write(&self, transaction: DBTransaction) -> io::Result<()> {
```

Comment:
We should rather use `enum` instead of integers for columns, to enforce the type safety and make sure we don't accidentally pass invalid integer into column.

So enum would look like:
```
pub enum DBCol {
  ColBlock,
  ColBlockMisc,
  ...
}
```
Then we would implement `From<DBCols>` for `&str` where we would manually map them to `&str`:
```
match self {
  ColBlock => "col_block",
....
}
```

Then we would store `HashMap<DBCol, *const ColumnFamily>` instead of `Vec<*const ColumnFamily>`. This will also allow us to get rid of `pub const COL_*` constants. Then `Store` method would have to take `column: DBCol` instead of the integer. 

The rationale is that it is currently not clear how much of the performance boost we get by not having strongly-typed columns, and store them in the vector. Unless we have a benchmark that shows that there is a performance difference I would rather use the strongly-typed version. (You could actually measure it by running https://github.com/nearprotocol/nearcore/tree/master/runtime/runtime-params-estimator and checking the storage operation fees). Then even if we see the difference we can still do the following:
```
repr(usize)
pub enum DBCol {
  ColBlock,
  ColBlockMisc,
  ...
}
```
and keep `Vec<*const ColumnFamily>`. But for now let's use `HashMap<DBCol, *const ColumnFamily>`

Reply:
Makes sense. Only thing is would HashMap slightly less efficient than Vector? If get is often this can make a difference

Reply:
Found we can use enum as C-like to get both type safety and (the unknown) performance bonus of enum:
https://doc.rust-lang.org/rust-by-example/custom_types/enum/c_like.html
so the interface with db.get would be `fn get(&self, col DBCol, key)`, and impl I can still use vector `db.get_cf(col as usize)`

Reply:
> Found we can use enum as C-like to get both type safety and (the unknown) performance bonus of enum:

That's what `repr(usize)` or `repr(u8)` accomplishes.

Reply:
But "C-like" `enum DBCol` is still enum, and use `fn get(&self, col DBCol, key), rustc will refuse any col that is not a member of DBCol, so we get type safety here